test/jdk/sun/security/pkcs11/Cipher/TestCipherTextStealingMultipart.java

```diff
@@ -0,0 +1,259 @@
+        }
+        return plaintextChunks;
+    }
+
+    private static byte[] computeExpected(byte[] jointPlaintext)
+            throws Exception {
+        byte[] ciphertext = sunJCECipher.doFinal(jointPlaintext);
+        if (ciphertext.length != jointPlaintext.length) {
+            throw new Exception("In CTS mode, ciphertext and plaintext should" +
+                    " have the same length. However, SunJCE's CTS cipher " +
+                    "returned a ciphertext of " + ciphertext.length + " bytes" +
+                    " and plaintext has " + jointPlaintext.length + " bytes.");
+        }
+        return ciphertext;
+    }
+
+    private enum CheckType {CIPHERTEXT, PLAINTEXT}
+
+    private enum OutputType {BYTE_ARRAY, DIRECT_BYTE_BUFFER}
+
+    private static void check(CheckType checkType, OutputType outputType,
+            byte[] expected, ByteBuffer actualBuf) throws Exception {
+        byte[] actual;
+        if (actualBuf.hasArray()) {
+            actual = actualBuf.array();
+        } else {
+            actual = new byte[actualBuf.position()];
+            actualBuf.position(0).get(actual);
+        }
+        if (!Arrays.equals(actual, expected)) {
+            throw new Exception("After " + switch (checkType) {
+                case CIPHERTEXT -> "encrypting";
+                case PLAINTEXT -> "decrypting";
+            } + " into a " + switch (outputType) {
+                case BYTE_ARRAY -> "byte[]";
+                case DIRECT_BYTE_BUFFER -> "direct ByteBuffer";
+            } + ", " + checkType.name().toLowerCase() + "s don't match:" + LF +
+                    "  Expected: " + repr(expected) + LF +
+                    "    Actual: " + repr(actual));
+        }
+    }
+
+    private static void doMultipart(int... chunkSizes) throws Exception {
+        int totalLength = IntStream.of(chunkSizes).sum();
+        byte[][] plaintextChunks = generateChunks(totalLength, chunkSizes);
+
+        ByteBuffer jointPlaintextBuf = ByteBuffer.allocate(totalLength);
+        for (byte[] plaintextChunk : plaintextChunks) {
+            jointPlaintextBuf.put(plaintextChunk);
+        }
+        byte[] jointPlaintext = jointPlaintextBuf.array();
+        byte[] expectedCiphertext = computeExpected(jointPlaintext);
+
+        // Check that the output array offset does not affect the penultimate
+        // block length calculation.
+        int outOfs = 1;
+
+        Cipher cipher = Cipher.getInstance(ALGORITHM, sunPKCS11);
+
+        // Encryption test, with byte[]
+        cipher.init(Cipher.ENCRYPT_MODE, KEY, IV);
+        ByteBuffer actualCiphertextBuf = ByteBuffer.allocate(totalLength);
+        for (byte[] plaintextChunk : plaintextChunks) {
+            actualCiphertextBuf.put(cipher.update(plaintextChunk));
+        }
+
+        byte [] outArray = new byte[cipher.getOutputSize(0) + outOfs];
+        cipher.doFinal(outArray, outOfs);
+        actualCiphertextBuf.put(outArray, outOfs, outArray.length - outOfs);
```

Comment:
Add this offset testing to the decryption part as well?

Reply:
Ok, I'll add it but it's not the same for decryption because it's the cipher text what has (potentially) to be reordered and output buffer will be for plain text. In fact, the reordering is over `padBuffer` (that has the last bytes of cipher text) for decryption.

Reply:
Thanks, yes, it's not about testing the byte-swapping part, but rather just to make sure the specified offset is correctly handled.